source/common/router/router.cc

```diff
@@ -423,20 +441,25 @@ Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_strea
     do_shadowing_ = false;
   }
 
-  if (buffering) {
-    // If we are going to buffer for retries or shadowing, we need to make a copy before encoding
-    // since it's all moves from here on.
-    Buffer::OwnedImpl copy(data);
-    upstream_request_->encodeData(copy, end_stream);
+  for (auto& upstream_request : upstream_requests_) {
+    if (buffering) {
+      // We need to make a copy before encoding since it's all moves from here
+      // on if we might have multiple upstream requests or traffic
+      // shadowing/retries.
+      Buffer::OwnedImpl copy(data);
+      upstream_request->encodeData(copy, end_stream);
+    } else {
+      upstream_request->encodeData(data, end_stream);
```

Comment:
AFIK on this path the common case is void Filter::UpstreamRequest::encodeData -> request_encoder_->encodeData which is allowed to move or drain data.   I'm not sure both request bodies will make it upstream.

I strongly encourage some integration tests here.  Even if the internals of router/hcm change I think the overall data flow between Envoy and upstream should remain the same, so I don't think you'll end up with any time wasted!

Reply:
+1 I don't think this will work if there are multiple upstreams without copying all but on of them. Is this path tested?

Reply:
I think I should revert this code and instead do an assert that `upstream_requests_.size() == 1`